File: sdk/csharp/src/Http/HttpClient.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using MrGaming.SDK.Exceptions;
using MrGaming.SDK.Models;

namespace MrGaming.SDK.Http
{
    /// <summary>
    /// HTTP客户端配置
    /// </summary>
    public class HttpClientConfig
    {
        public string BaseUrl { get; set; } = string.Empty;
        public TimeSpan Timeout { get; set; } = TimeSpan.FromSeconds(30);
        public int RetryCount { get; set; } = 3;
        public TimeSpan RetryDelay { get; set; } = TimeSpan.FromSeconds(1);
        public Dictionary<string, string> DefaultHeaders { get; set; } = new Dictionary<string, string>();
    }

    /// <summary>
    /// HTTP客户端
    /// </summary>
    public class HttpClient : IDisposable
    {
        private readonly System.Net.Http.HttpClient _httpClient;
        private readonly HttpClientConfig _config;
        private readonly ILogger<HttpClient>? _logger;
        private string? _authToken;
        private bool _disposed;

        public HttpClient(HttpClientConfig config, ILogger<HttpClient>? logger = null)
        {
            _config = config ?? throw new ArgumentNullException(nameof(config));
            _logger = logger;

            _httpClient = new System.Net.Http.HttpClient
            {
                Timeout = config.Timeout,
                BaseAddress = new Uri(config.BaseUrl)
            };

            // 设置默认请求头
            _httpClient.DefaultRequestHeaders.Add("User-Agent", "MrGaming.SDK.CSharp/1.0.0");
            _httpClient.DefaultRequestHeaders.Add("X-SDK-Version", "1.0.0");
            _httpClient.DefaultRequestHeaders.Add("X-SDK-Platform", "C#");

            foreach (var header in config.DefaultHeaders)
            {
                _httpClient.DefaultRequestHeaders.TryAddWithoutValidation(header.Key, header.Value);
            }
        }

        /// <summary>
        /// 设置认证令牌
        /// </summary>
        public void SetAuthToken(string token)
        {
            if (string.IsNullOrWhiteSpace(token))
            {
                throw new ArgumentException("认证令牌不能为空", nameof(token));
            }

            _authToken = token.Trim();
            _httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", _authToken);
        }

        /// <summary>
        /// 清除认证令牌
        /// </summary>
        public void ClearAuthToken()
        {
            _authToken = null;
            _httpClient.DefaultRequestHeaders.Authorization = null;
        }

        /// <summary>
        /// 设置自定义请求头
        /// </summary>
        public void SetHeader(string name, string value)
        {
            if (string.IsNullOrWhiteSpace(name))
            {
                throw new ArgumentException("请求头名称不能为空", nameof(name));
            }

            if (string.IsNullOrWhiteSpace(value))
            {
                _httpClient.DefaultRequestHeaders.Remove(name);
            }
            else
            {
                _httpClient.DefaultRequestHeaders.TryAddWithoutValidation(name, value);
            }
        }

        /// <summary>
        /// GET请求
        /// </summary>
        public async Task<ApiResponse<T>> GetAsync<T>(string endpoint, Dictionary<string, string>? parameters = null, CancellationToken cancellationToken = default)
        {
            var url = BuildUrl(endpoint, parameters);
            return await ExecuteRequestAsync<ApiResponse<T>>(HttpMethod.Get, url, null, cancellationToken);
        }

        /// <summary>
        /// POST请求
        /// </summary>
        public async Task<ApiResponse<T>> PostAsync<T>(string endpoint, object? data = null, CancellationToken cancellationToken = default)
        {
            return await ExecuteRequestAsync<ApiResponse<T>>(HttpMethod.Post, endpoint, data, cancellationToken);
        }

        /// <summary>
        /// PUT请求
        /// </summary>
        public async Task<ApiResponse<T>> PutAsync<T>(string endpoint, object? data = null, CancellationToken cancellationToken = default)
        {
            return await ExecuteRequestAsync<ApiResponse<T>>(HttpMethod.Put, endpoint, data, cancellationToken);
        }

        /// <summary>
        /// PATCH请求
        /// </summary>
        public async Task<ApiResponse<T>> PatchAsync<T>(string endpoint, object? data = null, CancellationToken cancellationToken = default)
        {
            return await ExecuteRequestAsync<ApiResponse<T>>(HttpMethod.Patch, endpoint, data, cancellationToken);
        }

        /// <summary>
        /// DELETE请求
        /// </summary>
        public async Task<ApiResponse<T>> DeleteAsync<T>(string endpoint, CancellationToken cancellationToken = default)
        {
            return await ExecuteRequestAsync<ApiResponse<T>>(HttpMethod.Delete, endpoint, null, cancellationToken);
        }

        /// <summary>
        /// 执行HTTP请求
        /// </summary>
        private async Task<TResponse> ExecuteRequestAsync<TResponse>(HttpMethod method, string endpoint, object? data, CancellationToken cancellationToken)
        {
            var attempt = 0;
            Exception? lastException = null;

            while (attempt < _config.RetryCount)
            {
                attempt++;
                try
                {
                    return await ExecuteSingleRequestAsync<TResponse>(method, endpoint, data, cancellationToken);
                }
                catch (Exception ex) when (ShouldRetry(ex) && attempt < _config.RetryCount)
                {
                    lastException = ex;
                    _logger?.LogWarning(ex, "请求失败，{delay}ms后进行第{attempt}次重试: {url}",
                        _config.RetryDelay.TotalMilliseconds, attempt + 1, endpoint);

                    await Task.Delay(_config.RetryDelay, cancellationToken);
                }
            }

            throw lastException ?? new InvalidOperationException("未知错误");
        }

        /// <summary>
        /// 执行单次HTTP请求
        /// </summary>
        private async Task<TResponse> ExecuteSingleRequestAsync<TResponse>(HttpMethod method, string endpoint, object? data, CancellationToken cancellationToken)
        {
            using var request = new HttpRequestMessage(method, endpoint);

            if (data != null)
            {
                var json = JsonSerializer.Serialize(data, new JsonSerializerOptions
                {
                    PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
                    DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull
                });

                request.Content = new StringContent(json, Encoding.UTF8, "application/json");
            }

            _logger?.LogDebug("发送HTTP请求: {method} {url}", method.Method, endpoint);
            if (data != null)
            {
                _logger?.LogDebug("请求内容: {content}", await request.Content!.ReadAsStringAsync(cancellationToken));
            }

            using var response = await _httpClient.SendAsync(request, cancellationToken);
            var responseContent = await response.Content.ReadAsStringAsync(cancellationToken);

            _logger?.LogDebug("收到HTTP响应: {status} {url}", response.StatusCode, endpoint);
            _logger?.LogDebug("响应内容: {content}", responseContent);

            if (!response.IsSuccessStatusCode)
            {
                throw CreateApiException(response, responseContent);
            }

            try
            {
                var result = JsonSerializer.Deserialize<TResponse>(responseContent, new JsonSerializerOptions
                {
                    PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
                    PropertyNameCaseInsensitive = true
                });

                return result ?? throw new InvalidOperationException("响应反序列化失败");
            }
            catch (JsonException ex)
            {
                _logger?.LogError(ex, "响应反序列化失败: {content}", responseContent);
                throw new ApiException("响应格式错误", (int)response.StatusCode, responseContent);
            }
        }

        /// <summary>
        /// 判断是否应该重试
        /// </summary>
        private bool ShouldRetry(Exception exception)
        {
            return exception is NetworkException ||
                   (exception is ApiException apiEx && IsRetryableStatusCode(apiEx.StatusCode));
        }

        /// <summary>
        /// 判断状态码是否可重试
        /// </summary>
        private static bool IsRetryableStatusCode(int? statusCode)
        {
            if (!statusCode.HasValue) return false;

            var retryableStatusCodes = new[]
            {
                408, // Request Timeout
                429, // Too Many Requests
                500, // Internal Server Error
                502, // Bad Gateway
                503, // Service Unavailable
                504, // Gateway Timeout
                520, // Unknown Error
                521, // Web Server Is Down
                522, // Connection Timed Out
                523, // Origin Is Unreachable
                524  // A Timeout Occurred
            };

            return retryableStatusCodes.Contains(statusCode.Value);
        }

        /// <summary>
        /// 创建API异常
        /// </summary>
        private ApiException CreateApiException(HttpResponseMessage response, string content)
        {
            var statusCode = (int)response.StatusCode;
            var message = $"HTTP {statusCode}: {response.ReasonPhrase}";

            try
            {
                var errorResponse = JsonSerializer.Deserialize<Dictionary<string, object>>(content);
                if (errorResponse?.ContainsKey("message") == true)
                {
                    message = errorResponse["message"].ToString() ?? message;
                }
            }
            catch
            {
                // 忽略JSON解析错误
            }

            return new ApiException(message, statusCode, content);
        }

        /// <summary>
        /// 构建URL
        /// </summary>
        private static string BuildUrl(string endpoint, Dictionary<string, string>? parameters)
        {
            if (parameters == null || parameters.Count == 0)
            {
                return endpoint;
            }

            var queryString = string.Join("&",
                parameters.Select(kvp => $"{Uri.EscapeDataString(kvp.Key)}={Uri.EscapeDataString(kvp.Value)}"));

            return string.IsNullOrEmpty(queryString) ? endpoint : $"{endpoint}?{queryString}";
        }

        /// <summary>
        /// 释放资源
        /// </summary>
        public void Dispose()
        {
            if (!_disposed)
            {
                _httpClient?.Dispose();
                _disposed = true;
            }
        }
    }
}

File: sdk/nodejs/src/http-client.ts
/**
 * HTTP客户端
 */

import axios, { AxiosInstance, AxiosRequestConfig, AxiosResponse, AxiosError } from 'axios';
import { ApiError, NetworkError } from './errors';
import { retry, delay } from './utils';
import { ApiResponse } from './types';

export interface HttpClientConfig {
  baseURL: string;
  timeout?: number;
  retryCount?: number;
  retryDelay?: number;
  headers?: Record<string, string>;
}

export class HttpClient {
  private axios: AxiosInstance;
  private retryCount: number;
  private retryDelay: number;

  constructor(config: HttpClientConfig) {
    this.retryCount = config.retryCount || 3;
    this.retryDelay = config.retryDelay || 1000;

    this.axios = axios.create({
      baseURL: config.baseURL,
      timeout: config.timeout || 30000,
      headers: {
        'Content-Type': 'application/json',
        'User-Agent': '@mr-gaming/sdk-nodejs/1.0.0',
        ...config.headers
      }
    });

    this.setupInterceptors();
  }

  /**
   * 设置请求和响应拦截器
   */
  private setupInterceptors(): void {
    // 请求拦截器
    this.axios.interceptors.request.use(
      (config) => {
        // 添加请求时间戳
        config.metadata = { startTime: Date.now() };

        console.log(`[HTTP Request] ${config.method?.toUpperCase()} ${config.url}`, {
          params: config.params,
      data: config.data
    });

    return config;
  },
  (error) => {
    console.error('[HTTP Request Error]', error);
    return Promise.reject(error);
  }
);

// 响应拦截器
this.axios.interceptors.response.use(
  (response) => {
    const duration = Date.now() - (response.config.metadata?.startTime || 0);

    console.log(`[HTTP Response] ${response.config.method?.toUpperCase()} ${response.config.url}`, {
      status: response.status,
      duration: `${duration}ms`,
      data: response.data
    });

    return response;
  },
  async (error: AxiosError) => {
    const config = error.config as any;
    const duration = Date.now() - (config?.metadata?.startTime || 0);

    console.error(`[HTTP Error] ${config?.method?.toUpperCase()} ${config?.url}`, {
      status: error.response?.status,
      duration: `${duration}ms`,
      message: error.message,
      data: error.response?.data
    });

    // 如果是网络错误并且设置了重试
    if (this.shouldRetry(error) && (!config._retryCount || config._retryCount < this.retryCount)) {
      config._retryCount = (config._retryCount || 0) + 1;

      console.log(`[HTTP Retry] Attempt ${config._retryCount}/${this.retryCount} for ${config?.method?.toUpperCase()} ${config?.url}`);

      // 指数退避策略
      const retryDelay = this.retryDelay * Math.pow(2, config._retryCount - 1);
      await delay(retryDelay);

      return this.axios(config);
    }

    return Promise.reject(this.handleError(error));
  }
);
}

/**
 * 判断是否应该重试
 */
private shouldRetry(error: AxiosError): boolean {
  // 网络错误
  if (!error.response) {
    return true;
  }

  const status = error.response.status;

  // 可重试的HTTP状态码
  const retryableStatuses = [
    408, // Request Timeout
    429, // Too Many Requests
    500, // Internal Server Error
    502, // Bad Gateway
    503, // Service Unavailable
    504, // Gateway Timeout
    520, // Unknown Error
    521, // Web Server Is Down
    522, // Connection Timed Out
    523, // Origin Is Unreachable
    524  // A Timeout Occurred
  ];

  return retryableStatuses.includes(status);
}

/**
 * 处理错误
 */
private handleError(error: AxiosError): Error {
  if (!error.response) {
    // 网络错误
    return new NetworkError(
      error.message || '网络连接失败',
      {
        code: error.code,
        message: error.message,
        config: error.config
      }
    );
  }

  const response = error.response;
  const data = response.data as any;

  // 构建错误消息
  let message = '请求失败';
  if (data?.message) {
    message = data.message;
  } else if (data?.error) {
    message = data.error;
  } else {
    message = `HTTP ${response.status}: ${response.statusText}`;
  }

  const apiError = new ApiError(
    message,
    response.status,
    data
  );

  // 添加额外的错误信息
  if (data?.code) {
    (apiError as any).errorCode = data.code;
  }

  if (data?.details) {
    (apiError as any).errorDetails = data.details;
  }

  return apiError;
}

/**
 * 设置认证token
 */
setAuthToken(token: string): void {
  this.axios.defaults.headers.common['Authorization'] = `Bearer ${token}`;
}

/**
 * 清除认证token
 */
clearAuthToken(): void {
  delete this.axios.defaults.headers.common['Authorization'];
}

/**
 * 设置自定义请求头
 */
setHeader(key: string, value: string): void {
  this.axios.defaults.headers.common[key] = value;
}

/**
 * GET请求
 */
async get<T = any>(
  url: string,
  params?: Record<string, any>,
  config?: AxiosRequestConfig
): Promise<ApiResponse<T>> {
  try {
    const response = await this.axios.get<ApiResponse<T>>(url, {
      ...config,
      params
    });

    return response.data;
  } catch (error) {
    throw error;
  }
}

/**
 * POST请求
 */
async post<T = any>(
  url: string,
  data?: any,
  config?: AxiosRequestConfig
): Promise<ApiResponse<T>> {
  try {
    const response = await this.axios.post<ApiResponse<T>>(url, data, config);
    return response.data;
  } catch (error) {
    throw error;
  }
}

/**
 * PUT请求
 */
async put<T = any>(
  url: string,
  data?: any,
  config?: AxiosRequestConfig
): Promise<ApiResponse<T>> {
  try {
    const response = await this.axios.put<ApiResponse<T>>(url, data, config);
    return response.data;
  } catch (error) {
    throw error;
  }
}

/**
 * PATCH请求
 */
async patch<T = any>(
  url: string,
  data?: any,
  config?: AxiosRequestConfig
): Promise<ApiResponse<T>> {
  try {
    const response = await this.axios.patch<ApiResponse<T>>(url, data, config);
    return response.data;
  } catch (error) {
    throw error;
  }
}

/**
 * DELETE请求
 */
async delete<T = any>(
  url: string,
  config?: AxiosRequestConfig
): Promise<ApiResponse<T>> {
  try {
    const response = await this.axios.delete<ApiResponse<T>>(url, config);
    return response.data;
  } catch (error) {
    throw error;
  }
}

/**
 * 文件上传
 */
async upload<T = any>(
  url: string,
  file: Buffer | Blob,
  fileName: string,
  additionalData?: Record<string, any>,
  config?: AxiosRequestConfig
): Promise<ApiResponse<T>> {
  const formData = new FormData();

  // 添加文件
  if (file instanceof Buffer) {
    formData.append('file', new Blob([file]), fileName);
  } else {
    formData.append('file', file, fileName);
  }

  // 添加额外数据
  if (additionalData) {
    for (const [key, value] of Object.entries(additionalData)) {
      formData.append(key, String(value));
    }
  }

  try {
    const response = await this.axios.post<ApiResponse<T>>(url, formData, {
      ...config,
      headers: {
        'Content-Type': 'multipart/form-data',
        ...config?.headers
      }
    });

    return response.data;
  } catch (error) {
    throw error;
  }
}
}